### PR TITLE
LakeFSFS - TokenProvider AWS (no login yet)

### LINKED ISSUE #7537: lakeFSFS - support External principal login with IAM role
Depends on #7536  and login flow to be implemented. 

**D.O.D**

- Extend our code [Hadoop client to support it](https://github.com/treeverse/lakeFS/blob/master/clients/hadoopfs/src/main/java/io/lakefs/LakeFSClient.java#L23) Pyspark configuration would be something like  `spark.hadoop.fs.lakefs.role.arn`.
- The client should be able to persist the token it gets in a file style `~/.lakefs/cli/cache`  - configurable. Not every environment can store tokens in a file. 
- The client should be able to detect if the token is about-to-expire and renew it on the fly (during initialization only for now). 

# High level flow of getting valid lakeFS JWT: 

1. Create a valid AWS Session, a valid session is defined as when calling `GetCallerIdentity`  the ARN returned is mapped in lakeFS as external principal. - **in the context of Spark / lakeFSFS / Hadoop that's a Credentials Provider**.
1. Use the session to generate a **presigned** GetCallerIdentity and compose an identity token. 
1. Do "/login" with lakeFS and receive a valid JWT token to work with. (lakeFS token i.e list repositories). 

# Credentials Provider - getting valid AWS session

That section is a deep dive on Step 1 - how does lakeFSFS get a valid AWS session? 

Similarly to how aws-hadoop use [providers](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html#Authenticating_with_S3) we are going to need few of these. 
Today we already have one implicitly, LakeFS Client that uses lakeFS AccessKey and SecretKey. 
Let's call it `BasicCredentialsProvider`. 

We are going to need 2 more Credential Providers and preparation for custom provider that will be implemented by our users **(The custom provider support is not a blocker for release)**.

1. [TemporaryAWSCredentialsProvider](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html#Using_Session_Credentials_with_TemporaryAWSCredentialsProvider): can be obtained from the Amazon Security Token Service; these consist of an access key, a secret key, and a session token.
Set the session key in the property `fs.lakefs.session.token`, and the `access` and `secret key` properties to those of this temporary session.

Pros: the most flexible way, user provide us with the credentials we need. 

Cons: Renewal of the lakeFS JWT is not possible out of the box since the session creds are temporary and the next time renewing will be probably expired. 

2. [InstanceProfileCredentialsProvider](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html#EC2_IAM_Metadata_Authentication_with_InstanceProfileCredentialsProvider). 

Applications running in EC2 may associate an IAM role with the VM and query the [EC2 Instance Metadata Service](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) for credentials to access S3. Within the AWS SDK, this functionality is provided by InstanceProfileCredentialsProvider, which internally enforces a singleton instance in order to prevent throttling problem.

Pros: We can control the renewal of lakeFS token since we can always create the session ourselves. Common use-case in environments like Databricks cluster. 

Cons: Every spark job running on those instances will share the same lakeFS user. 

## Extensions: 

1. [Custom Provider](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html#Custom_AWS_Credential_Providers_and_Apache_Spark) - not a blocker for the current release but important to design based on that - allow customers pass their own credentials provider and use it in the environment. 

1. [AssumedRoleCredentialProvider](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/assumed_roles.html ) - Allow users to pass specific arn to assume in order to authenticate with lakeFS. 
The challenge is how do we authentication to AWS in order to create a valid session that will allow us to assume the role? 
In `aws-hadoop` that provider requires additional long-lived IAM credentials AccessKey/SecretKey to assume that role. 
That's long-lived credentials dependency is not a very good use-case for us. 
We might implement this and / or support using other providers we have such as `InstanceProfileCredentialsProvider`  to get a valid session and then call AssumeRole.

### PR DESCRIPTION
**Pull request by Isan-Rivkin**
Closes #7537 (without call to login, waiting on #7578). 
No breaking changes. Currently not document and internal only.

### CODE REVIEW

**Comment by Isan-Rivkin** [comment]
@arielshaqed it's still a draft but please take a look at the draft im sure you will have comments.
If it's not clear then i'll come back and continue.

File: clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java
Comment (arielshaqed) [inline clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:186]: I don't understand, it looks like this will _never_ refresh!  The first time it calls newToken it sets lakeFSAuthToken.  And every time after that !needsNewToken, so it does nothing.

**Review by arielshaqed — COMMENTED** [review]
THANKS!

Very nice.  All code here is very good.  But I think I am missing several parts, notably I don't understand how refresh can ever **re--**fresh the token, it will only set it once.  I think we will need an additional executor / sleeping thread / something that periodically refreshes.  So not sure what parts are included in this draft, basically.

AFAICT this requires sigV4 on the AWS endpoint.  This is probably cool, but best validate versus requirements.  In particular, us-east-1 is always special with regards to sigV4/sigV2.

File: clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java
Comment (Isan-Rivkin) [reply clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:186]: I can see very clearly see why it's confusing sorry for that. 

1. Regardless of implementation which we can discuss - the goal of this function is to create a new token (with new TTL) if possible. 
1. I'm trying to make a standard for calling refresh token, mimicking how [AWSCredentialProvider](https://github.com/aws/aws-sdk-java/blob/1.7.4/src/main/java/com/amazonaws/auth/AWSCredentialsProvider.java#L44) did that.
1. I'm open for discussion on removing it for the initial release - e.g no refresh will be delivered.

File: clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java
Comment (Isan-Rivkin) [reply clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:16]: It does unless the class is not declared as `public` 👀

File: clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java
Comment (Isan-Rivkin) [reply clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:55]: removed

File: clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java
Comment (Isan-Rivkin) [reply clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:16]: THANK YOU this is unnecessary, removed it and adopted the code.

File: clients/hadoopfs/src/main/java/io/lakefs/Constants.java
Comment (Isan-Rivkin) [reply clients/hadoopfs/src/main/java/io/lakefs/Constants.java]: fixed duration. 
will add documentation **before** merging this as a final step in additional PR into this one, i just want to be sure about the actual configuration

File: clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java
Comment (Isan-Rivkin) [reply clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java]: fixed

**Comment by Isan-Rivkin** [comment]
@arielshaqed please re-review. 
I have addressed all your comments.
This PR will require additional look before final approve it depends on missing functionality PR #7578  being merged. 
Once that merged I'll use the calls in `newToken()`.

File: clients/hadoopfs/pom.xml
Comment (arielshaqed) [inline clients/hadoopfs/pom.xml:290]: Why +2?

File: clients/hadoopfs/src/main/java/io/lakefs/LakeFSClient.java
Comment (arielshaqed) [inline clients/hadoopfs/src/main/java/io/lakefs/LakeFSClient.java:47]: nit:
```suggestion
        } else {
```

File: clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java
Comment (arielshaqed) [inline clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:33]: This is a very odd constructor.  It does nothing except change the scope.  Can we get rid of it?

File: clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java
Comment (arielshaqed) [reply clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:142]: So please document what it does: it clearly confuses at least one reader of the code.

File: clients/hadoopfs/src/main/java/io/lakefs/auth/GeneratePresignGetCallerIdentityRequest.java
Comment (arielshaqed) [inline clients/hadoopfs/src/main/java/io/lakefs/auth/GeneratePresignGetCallerIdentityRequest.java:1]: Nit: this is a value type.  Could we use [org.immutables.value](https://immutables.github.io/)?

File: clients/hadoopfs/src/main/java/io/lakefs/auth/LakeFSTokenProvider.java
Comment (arielshaqed) [inline clients/hadoopfs/src/main/java/io/lakefs/auth/LakeFSTokenProvider.java:8]: Please document interface methods.

File: clients/hadoopfs/src/test/java/io/lakefs/FSConfigurationTest.java
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/FSConfigurationTest.java:59]: ?

File: clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:41]: This needs to be configurable.  Think AWS in China, govclouds, various regional endpoints, etc.

File: clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:43]: Can we use e.g. [ImmutableMap](https://guava.dev/releases/23.0/api/docs/com/google/common/collect/ImmutableMap.html)?

File: clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:54]: Please position "+" signs consistently.

File: clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:59]: Ideally you would extract the credential to a group using `(\w+)`, match, and then test equality to the AWS access key ID.  The current code might behave weirdly if the access key ID is somehow invalid and contains strange characters.  We can work around that, or we can make an effort to explain why it cannot happen, or we can make the effort to quote the access key ID -- but it will still be simplest simply not to use a regexp to do that.

I know it's just a test.  OTOH, "code returned a bad access key" would be a really good error message :-)

File: clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:64]: This unit test verifies that the code returns all query parameters in a particular order.  But I do not think that this should be part of the requirement for what we are testing.

So I would really prefer to _parse_ the HRL, and then verify that the fields are correct.

**Review by arielshaqed — APPROVED** [review]
Neat!

How was this tested?  Can we incorporate it into an Esti test, or into the contract test?  I worry about regressing.

File: clients/hadoopfs/pom.xml
Comment (Isan-Rivkin) [reply clients/hadoopfs/pom.xml:290]: 1.14 + contains all external principal API's. 
So it will end up 1.14.x with the login API merged.

File: clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java
Comment (Isan-Rivkin) [reply clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java:33]: That's a default constructor I'm using it because the child class(es) `TemporaryAWSCredentialsLakeFSTokenProvider` initiates it and they also do some work before initiating it.

File: clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java
Comment (Isan-Rivkin) [reply clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:43]: Dependency that exists only in tests we discussed and decided not to introduce into lakeFSFS.

File: clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java
Comment (Isan-Rivkin) [reply clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:41]: Not sure what you mean? this **is** configurable, it's just a test Im initiating the address, I could put other addresses.

File: clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java
Comment (Isan-Rivkin) [reply clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:64]: updated the test to check and compare each parameter one by one as suggested, thanks!

File: clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java
Comment (Isan-Rivkin) [reply clients/hadoopfs/src/test/java/io/lakefs/auth/GetCallerIdentityV4PresignerTest.java:59]: updated with specific check for that param and a message

**Comment by Isan-Rivkin** [comment]
Thanks @arielshaqed addressed all your comments. 
Merging this PR and waiting for #7578 to be merged. 
Once it's merged I'll create additional PR with call to that API in the `LakeFSClient` and documentation accordingly.

**Pull request merged**
